### DISCUSSION TITLE
Start writing tests

### DISCUSSION THREAD

**Discussion by bdm-k** [artisticat1/obsidian-latex-suite#293] [General]
@artisticat1 
Writing tests every time you fix a bug or add a new feature is a good practice to ensure that the behavior doesn't change unintentionally. Could you please introduce a framework for writing tests? I found some articles about automated tests of obsidian plugins [here](https://publish.obsidian.md/hub/04+-+Guides%2C+Workflows%2C+%26+Courses/for+Plugin+Developers+to+Automate+Tests).

**Comment by artisticat1** [comment]
Yes, adding tests is something I've been wanting to do for a while but haven't found the time to do.

https://github.com/obsidianmd/obsidian-api/issues/13#issuecomment-1003880942 seems helpful, in case anyone is interested in developing a PR.

**Comment by bdm-k** [comment]
I introduced Jest as the testing framework and attempted to write a test. I found that since we import resources from the obsidian package in "editor_utils.ts", it is impossible to test functions that depend on the resources in "editor_utils.ts". That is, we are unable to test the function `onKeydown`, for example.

**Comment by bdm-k** [comment]
I leave the way to setup Jest here:

1. Install the following packages: jest, jest-environment-jsdom, @types/jest, ts-jest.
2. Edit "jest.config.js" as follows:
```javascript
/** @type {import('ts-jest').JestConfigWithTsJest} **/
module.exports = {
  testEnvironment: "jest-environment-jsdom",
  transform: {
    "^.+.tsx?$": ["ts-jest",{}],
  },
  moduleNameMapper: {
    "^src/(.+)$": "<rootDir>/src/$1",
  },
};
```

**Comment by artisticat1** [comment]
Thanks for taking a look!

> I found that since we import resources from the obsidian package in "editor_utils.ts", it is impossible to test functions that depend on the resources in "editor_utils.ts".

I think we can remove the references to Workspace and MarkdownView in "editor_utils.ts" if we remove the `iterateCM6()` function, like you've done in #322.

That just leaves the reference to Platform in `if (Platform.isMobile) return;`. Maybe we can isolate it somehow? (For context, I added this check to get the plugin working on iOS. When omitted, the plugin would break down and produce erratic behaviour. There have been a lot of updates since then though, so maybe the check isn't required anymore?)
